verification/tpm2_test.go

```diff
@@ -0,0 +1,230 @@
+package verification
+
+import (
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"math/big"
+	"os"
+	"testing"
+
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm/legacy/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var (
+	tpmPath = flag.String("tpm-path", "/dev/tpm1", "Path to the TPM device (character device or a Unix socket).")
+	flush   = flag.String("flush", "all", "Flush existing handles")
+
+	handleNames = map[string][]tpm2.HandleType{
+		"all":       []tpm2.HandleType{tpm2.HandleTypeLoadedSession, tpm2.HandleTypeSavedSession, tpm2.HandleTypeTransient},
+		"loaded":    []tpm2.HandleType{tpm2.HandleTypeLoadedSession},
+		"saved":     []tpm2.HandleType{tpm2.HandleTypeSavedSession},
+		"transient": []tpm2.HandleType{tpm2.HandleTypeTransient},
+	}
+)
+
+func TestTPME2E(t *testing.T) {
+	supportNeeded := []string{"AutoInit", "X509"}
+	instance, err := GetTestTarget(supportNeeded)
+	if err != nil {
+		if err.Error() == "Requested support is not supported in the emulator" {
+			t.Skipf("[WARNING]: Failed executing TestSign command due to unsupported request. Hence, skipping the command execution")
+		} else {
+			t.Fatalf("[FATAL]: %s", err.Error())
+		}
+	}
+	testTPME2E(instance, t, false)
+}
+
+func getTpmNonce() (io.ReadWriteCloser, tpmutil.Handle, []byte, error) {
+
+	rwc, err := tpm2.OpenTPM(*tpmPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Can't open TPM %s: %v", *tpmPath, err)
+	}
+
+	totalHandles := 0
+	for _, handleType := range handleNames[*flush] {
+		handles, err := client.Handles(rwc, handleType)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error getting handles %s: %v", *tpmPath, err)
+			os.Exit(1)
+		}
+		for _, handle := range handles {
+			if err = tpm2.FlushContext(rwc, handle); err != nil {
+				fmt.Fprintf(os.Stderr, "Error flushing handle 0x%x: %v\n", handle, err)
+				os.Exit(1)
+			}
+			fmt.Printf("Handle 0x%x flushed\n", handle)
+			totalHandles++
+		}
+	}
+
+	sessHandle, nonce, err := tpm2.StartAuthSession(rwc, tpm2.HandleNull, tpm2.HandleNull, make([]byte, 16), nil, tpm2.SessionPolicy, tpm2.AlgNull, tpm2.AlgSHA256)
+	if err != nil {
+		fmt.Printf("StartAuthSession() failed: %v", err)
+		os.Exit(1)
+	}
+
+	return rwc, sessHandle, nonce, nil
+}
+
+func testTPME2E(d TestDPEInstance, t *testing.T, use_simulation bool) {
```

Comment:
General comment that it would be good to break this function up into smaller functions.

Reply:
Have broken down to 
startTpmSession()
getDigest() - creates digest to be passed to TPM
extractPubKey()
getPubKeyHandle - gets a handle from TPM for pub key loaded into it
getEncodedSignature()